### ISSUE TITLE
passthrough CloseFile Problem

### ISSUE THREAD

**Issue by thinkport** [billziss-gh/winfsp#186]
## Question

_Please consider asking questions in the [WinFsp Google Group](https://groups.google.com/forum/#!forum/winfsp) instead. Before asking a question please also consult the [WinFsp Frequently Asked Questions](https://github.com/billziss-gh/winfsp/wiki/Frequently-Asked-Questions)._

Thankyou for you work, I have another problem about passthrough. The passthrough missing some CloseFile, The passthrough hold the file, so some software don't work well sometimes.

I test passthrough in win7x64 and win10x64.

**reproduce**
step 1: passthrough-x86.exe -p c:\ -m m:
step 2: open m:\a.docx with the winword.exe
step 3: close the winword.exe process
step 4: call CreateFile(c:\a.txt), dwShareMode = 0, CreateFile Will return INVALID_HANDLE_VALUE, GetLastError = 32

**Comment by billziss-gh** [comment]
When an application closes a file, the Windows OS may still keep the file open for a number of reasons (e.g. caching). Unfortunately passthrough uses simple file handles to track its files, so this often results in conflicts.

It is possible to use the `FlushAndPurgeOnCleanup` flag to reduce the occurrences of this problem. This will instruct the FSD to ask the OS to close files more aggressively.

The real issue here is that passthrough was designed as an example file system for a tutorial and does not handle many scenarios very well. I have contemplated for some time to write an "advanced" passthrough file system that solves some of these problems (at the expense of loss of simplicity).

So in a nutshell I accept that this is a problem with the passthrough file system, but I do not expect to change it (other than perhaps adding the `FlushAndPurgeOnCleanup` flag) because I want it to continue serving as a sample file system.

**Comment by thinkport** [comment]
You can see some CloseFile lost. It's easy find this problem when I reparse file to WinFSP.
![image](https://user-images.githubusercontent.com/12952910/44439432-1ead1e80-a5f6-11e8-97aa-2af4fbabd163.png)

Explorer pass IRP_MJ_CLOSE to I/O Manger, and I/O manager pass IRP to WinFSP, but WinFSP doesn't call passthrough to close file sometimes. So I guess this is WinFSP problem. But it's difficult to deubg this problem, beacuse it only happens sometimes.

**Comment by billziss-gh** [comment]
@thinkport again this is expected behavior. If you want to convince yourself download a utility called [RAMMap](https://docs.microsoft.com/en-us/sysinternals/downloads/rammap) and issue an "empty standby list" command. You will suddenly see all the "missing" CLOSE's.

**Comment by billziss-gh** [comment]
Also please use [FileSpy](http://www.zezula.net/en/fstools/filespy.html) instead of Process Explorer as it does a better job at tracking all IRP's.

**Comment by billziss-gh** [comment]
BTW, one of the reasons that Windows keeps files open for long periods is caching. Consider a program that constantly opens, reads/writes and closes the same file (a rather common pattern). If Windows keeps the file open (despite what the program thinks) it will have a cache-miss on the first read and subsequent I/O will be satisfied from cache (until the cache gets purged for whatever reason). The alternative would be to re-prime the cache on the first read after every open.

**Comment by thinkport** [comment]
Thankyou for your reply. I do more test, I think you are right, filesystem maybe close file later, whatever softwares can work well.

I think I found the wrong way, problem is not that passthrough hold the file. I reparse file to WinFSP, I can't save file on the Desktop in Windows 10, WinWord will delete file when I save file, but WinWord work well in other folder. Windows 7 work well, no matter where the file is.

![a](https://user-images.githubusercontent.com/12952910/44453406-8f226280-a62b-11e8-99b9-ebf6f3ef2880.png)
![b](https://user-images.githubusercontent.com/12952910/44453487-b6792f80-a62b-11e8-805d-62f3c7023063.png)
